### PR TITLE
builtin/getopts: implement getopts builtin

### PR DESCRIPTION
**Pull request by delthas**
https://web.archive.org/web/20180906014022/http://pubs.opengroup.org/onlinepubs/9699919799/utilities/getopts.html

- not sure how to get the "shell positional arguments" (in the cases the user does not pass any args to getopts apart from the optstring and the variable name), i used `state->argv` and `state->argc` but did not try it
- i think getopts returns exactly the same outputs when either an option-argument is missing and the optstring does not start with a colon, or an unknown option character was found. so to find out which case we're in, we have to iterate all the options in the optstring to know if the option is there (so the case is "missing an argument"), or not ("unknown option"). not very expensive but not perfect. (we need to know which case we're in b/c we need to set or unset a variable depending on the case)
- i have checked a lot of corner cases against bash, and compared their environment variables and stderr outputs, with commands like `getopts a: test -a`, `getopts :a: test -a`, `getopts a test -v`, `getopts :a test -v`. i think my implementation is correct wrt most corner cases but i may have forgotten a few of them

### CODE REVIEW

**Comment by emersion** [comment]
>not sure how to get the "shell positional arguments" 

Yeah, it indeed refers to `state->argv` and `state->argc`. You can try setting `argv[1..]` by using the command `set a b c`.

>we have to iterate all the options in the optstring to know if the option is there

This seems reasonable to me.

>i have checked a lot of corner cases against bash

It would be nice to add tests for those.

File: builtin/getopts.c [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		optv = state->argv;
+	}
+	
+	const char *optind_str = mrsh_env_get(state, "OPTIND", NULL);
+	if (optind_str == NULL) {
+		fprintf(stderr, "getopts: OPTIND is not defined\n");
+		return EXIT_FAILURE;
+	}
+	char *endptr;
+	long optind_int = strtol(optind_str, &endptr, 10);
```
Comment (emersion) [inline builtin/getopts.c]: This should probably be named `optind_long`

File: builtin/getopts.c [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			mrsh_env_set(state, "OPTARG", value, MRSH_VAR_ATTRIB_NONE);
+		} else if (argv[1][0] != ':') {
+			mrsh_env_unset(state, "OPTARG");
+		} else {
+			// either missing option-argument or unknown option character
+			// in the former case, unset OPTARG
+			// in the latter case, set OPTARG to optopt
+			bool opt_exists = false;
+			size_t len = strlen(argv[1]);
+			for (int i = 0; i < len; ++i) {
```
Comment (emersion) [inline builtin/getopts.c]: This can be a `size_t`

File: builtin/getopts.c [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	
+	char optind_fmt[16];
+	snprintf(optind_fmt, sizeof(optind_fmt), "%d", optind);
+	mrsh_env_set(state, "OPTIND", optind_fmt, MRSH_VAR_ATTRIB_NONE);
+	
+	if (optopt != 0) {
+		if (opt == ':') {
+			char value[] = {(char)optopt, '\0'};
+			mrsh_env_set(state, "OPTARG", value, MRSH_VAR_ATTRIB_NONE);
+		} else if (argv[1][0] != ':') {
```
Comment (emersion) [inline builtin/getopts.c]: Is it possible to add a variable `optstring = argv[1]`? This would make it easier to follow the logic here.

File: builtin/getopts.c [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		mrsh_env_set(state, "OPTARG", optarg, MRSH_VAR_ATTRIB_NONE);
+	} else {
+		mrsh_env_unset(state, "OPTARG");
+	}
+	
+	char value[] = {opt == -1 ? (char)'?' : (char)opt, '\0'};
+	mrsh_env_set(state, argv[2], value, MRSH_VAR_ATTRIB_NONE);
+	
+	if (opt == -1) {
+		return 1;
```
Comment (emersion) [inline builtin/getopts.c]: `EXIT_FAILURE`

**Comment by emersion** [comment]
Thanks!
